cpp/serve/request.h

```diff
@@ -0,0 +1,222 @@
+
+/*!
+ * \brief The user submitted text-generation request, which contains
+ * a list of multi-modal inputs and a set of sampling configuration
+ * parameters.
+ */
+class RequestNode : public Object {
+ public:
+  Array<Input> inputs;
+  /*! \brief JSON string containing temperature, top_p, repetition_penalty, max_gen_len, etc. */
+  SampleConfig sampling_params;
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("inputs", &inputs);
+    v->Visit("sampling_params", &sampling_params);
+  }
+
+  bool SEqualReduce(const RequestNode* other, tvm::SEqualReducer equal) const {
+    return equal(inputs, other->inputs) && equal(sampling_params, other->sampling_params);
+  }
+
+  void SHashReduce(tvm::SHashReducer hash_reduce) const {
+    hash_reduce(inputs);
+    hash_reduce(sampling_params);
+  }
+
+  static constexpr const char* _type_key = "mlc.serve.Request";
+  static constexpr const bool _type_has_method_sequal_reduce = true;
+  static constexpr const bool _type_has_method_shash_reduce = true;
+  TVM_DECLARE_BASE_OBJECT_INFO(RequestNode, Object);
+};
+
+class Request : public ObjectRef {
+ public:
+  explicit Request(Array<Input> inputs, String sampling_params_json);
+
+  TVM_DEFINE_OBJECT_REF_METHODS(Request, ObjectRef, RequestNode);
+};
+
+/****************** Request State ******************/
+
+/*!
+ * \brief The state of a request with regard to some single model.
+ * \details In MLC LLM, the serving engine may leverage multiple models
+ * to fulfill a user generation request (e.g., use speculation decoding).
+ * For each request, we isolate its states (e.g. the generated tokens)
+ * on each model. This is to say, we use RequestModelState to store
+ * the state of a user request on a single model (rather than all models).
+ */
+class RequestModelStateNode : public Object {
+ public:
+  /*!
+   * \brief The corresponding request id of this state.
+   * It is the **physical index** of the request in the running request queue.
+   * If the request is on hold (not in the running queue), the request id
+   * should be -1.
+   */
+  int request_id = -1;
+  /*! \brief The corresponding model id of this state. */
+  int model_id = -1;
+  /*!
+   * \brief The committed generated token ids. A token is "committed"
+   * means it will no longer be updated (or changed).
+   */
+  std::vector<int32_t> committed_tokens;
+  /*! \brief The list of inputs yet for the model to prefill. */
+  Array<Input> input_to_prefill;
+  /*!
+   * \brief The draft generated token ids, which are usually generated
+   * by "small" speculative models. These tokens will be fed to a "large"
+   * model to determine the final result of speculation.
+   */
+  std::vector<int32_t> draft_output_tokens;
+  /*!
+   * \brief The probability distribution on each position in the
+   * draft. We keep the distributions for stochastic sampling when merging
+   * speculations from multiple models.
+   */
+  std::vector<std::vector<float>> draft_output_prob_dist;
```

Comment:
NDArray?, likely we will place everything together, nested vector is not as efficient here. We can add comment that they are reserved for small models.

Reply:
Add note: we only need this value when we have multiple parallel drafts